File: leetcode/75_LeetCode_Question/Array/search_in_rotated_sorted_array/solve.py
def find_target_1(nums, target):
    try: return nums.index(target)
    except: return -1

def binary_search_rotated(nums, target):
    r,l = len(nums) -1, 0
    while l <= r:
        mid = l + (r-l)//2
        if nums[mid] == target: return mid
        if nums[mid] >= nums[l]:
            if target > nums[mid] or target < nums[l]:
                l = mid + 1
            else: 
                r = mid - 1
        else:
            if target < nums[mid] or target > nums[r]:
                r = mid - 1
            else:
                l = mid + 1
    return -1

test_arr = [4,5,6,7,8,0,1,2]
test_target = 1
print(find_target_1(test_arr, test_target))
print(binary_search_rotated(test_arr, test_target))

File: leetcode/75_LeetCode_Question/Array/search_in_rotated_sorted_array/solve.rs
fn search_rotated_1(nums: &Vec<i32>, target: i32) -> i32 {
    for idx in 0..nums.len(){
        if nums[idx] == target {
            return idx as i32;
        }
    };
    return -1;
}

fn binary_search_rotated(nums: &Vec<i32>, target: i32) -> i32 {
    let mut r = nums.len()-1;
    let mut l = 0;
    let mut mid: usize;
    while l <= r {
        mid = l + (r-l)/2;
        if target == nums[mid]{
            return mid as i32;
        };

        if nums[mid] >= nums[l] {
            if target > nums[mid] || target < nums[l]{
                l = mid + 1;
            }else{
                r = mid - 1;
            }
        }else{
            if target < nums[mid] || target > nums[r]{
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
    }
    return -1;
}

fn main(){
    let test_vec = vec![4,5,6,7,8,9,10,1,2,3];
    let test_target = 1;
    println!("search_rotated_1: {:?}", search_rotated_1(&test_vec, test_target));
    println!("binary_search_rotated: {:?}", binary_search_rotated(&test_vec, test_target));
}
